### PR TITLE
[cuda] Remove host sync from masked_fill_ with a device value tensor

### PR DESCRIPTION
**Pull request by dumko2001**
## Summary

`masked_fill_` with a value that lives on the GPU is slower than it needs to be, and it cannot be captured into a CUDA graph.

The CUDA overload reads the fill value with `value.item()`:

```cpp
return masked_fill__cuda(self, mask, value.item());
```

`value.item()` copies one element from the GPU back to the host and blocks the calling thread until the copy lands. So every call with a device value stalls the host. The copy is also illegal during CUDA graph capture, so the op cannot be captured, and it forces a break in `torch.compile` cudagraph regions. You hit this when the fill value came from an earlier GPU computation, for example `x.masked_fill_(mask, scale)` where `scale` is a reduction result still on the device.

This PR reads the value on the GPU instead. For a device value it builds a `TensorIterator` over `self`, the broadcast mask, and the 0-dim value, and a small kernel selects the value where the mask is set. The 0-dim value broadcasts over the output, so there is no host round-trip and the op can be captured.

A CPU value never needed a sync, since `value.item()` there just reads host memory. So a CPU value keeps the old, cheaper path. Only a device value takes the new path, which is the case that used to stall.

The value is cast with `value.to(self.options())` before the kernel runs, so the result matches the existing scalar overload across the supported dtypes. The device, dtype, and overlap checks are preserved.

This follows recent host-sync removals in the CUDA backend, for example #186595 and #189305.

## Test Plan

```
python test/test_torch.py -k test_masked_fill_tensor_value
python test/test_torch.py -k test_masked_fill_tensor_value_cuda_graph
```

I also checked the change on an NVIDIA RTX A4000 (CUDA 12.1). I compiled the exact device-value path as an extension and compared it against the stock op:

- the device-value result matches the scalar overload for float32, float64, float16, bfloat16, and int64
- the stock `masked_fill_(mask, device_value)` fails inside `torch.cuda.graph` capture, while the on-device path captures and replays with the correct result

`lintrunner` is clean on the changed files.

This change was written with help from an AI assistant. I have read the diff and can explain every line.

### CODE REVIEW

File: test/test_torch.py [hunk 1]
```diff
@@ -3950,6 +3913,42 @@ def test_masked_fill_bool_tensor(self, device):
...
+        ref = dst.clone().masked_fill_(mask, 3.5)
+        val = torch.tensor(3.5, device=device)
+        out = dst.clone().masked_fill_(mask, val)
+        self.assertEqual(out, ref, atol=0, rtol=0)
+
+        # A host-side value tensor keeps working too.
+        out_cpu_val = dst.clone().masked_fill_(mask, torch.tensor(3.5))
+        self.assertEqual(out_cpu_val, ref, atol=0, rtol=0)
+
+    @onlyCUDA
```
Comment (Skylion007) [inline test/test_torch.py:3979]: Is there a better decorator for cudagraph support? MPS is getting its on graph support soon.

File: aten/src/ATen/native/cuda/Indexing.cu [hunk 2]
```diff
@@ -1833,7 +1846,41 @@ Tensor & masked_fill__cuda(Tensor& self, const Tensor & mask, const Tensor & val
   // `mask` to be CPU tensor. Check for `self` and `mask` being on same device
   // exists in `masked_fill__cuda` (Scalar version).
   TORCH_CHECK(!self.device().is_cpu(), "masked_fill_: Expected inputs to be on same device")
-  return masked_fill__cuda(self, mask, value.item());
+  // A host-side value carries no device sync, so keep the cheap Scalar path.
+  if (value.device().is_cpu()) {
+    return masked_fill__cuda(self, mask, value.item());
+  }
+  TORCH_CHECK(self.device() == mask.device(), "expected self and mask to be on the same device, but got mask on ",
```
Comment (Skylion007) [inline aten/src/ATen/native/cuda/Indexing.cu:1853]: If these aren't defined yet, they should be CHECK_VALUE, is these asserts already exist in other code paths, then keep it.

File: test/test_torch.py (hunk 1 above)
Comment (dumko2001) [reply test/test_torch.py:3979]: No backend-agnostic graph decorator exists yet, and this test calls torch.cuda.CUDAGraph/torch.cuda.graph directly, so @onlyCUDA is the right fit today. Cross-backend correctness is already covered by test_masked_fill_tensor_value — this test only adds the capture/replay check. Glad to add an MPS graph test once MPS capture lands.

File: aten/src/ATen/native/cuda/Indexing.cu (hunk 2 above)
Comment (dumko2001) [reply aten/src/ATen/native/cuda/Indexing.cu:1853]: hese are copied straight from the existing Scalar masked_fill__cuda overload (and match masked_fill_impl_cpu), which both use TORCH_CHECK. Keeping TORCH_CHECK for consistency rather than diverging a single copy to TORCH_CHECK_VALUE

File: aten/src/ATen/native/cuda/Indexing.cu [hunk 3]
```diff
@@ -1833,7 +1846,41 @@ Tensor & masked_fill__cuda(Tensor& self, const Tensor & mask, const Tensor & val
...
+    mask.device(), " and self on ", self.device());
+  TORCH_CHECK(mask.scalar_type() == kBool,
+    "masked_fill only supports boolean masks, but got dtype ", mask.scalar_type());
+  if (at::has_internal_overlap(self) == MemOverlap::Yes) {
+    TORCH_WARN(
+      "Use of masked_fill_ on expanded tensors is deprecated. "
```
Comment (albanD) [inline aten/src/ATen/native/cuda/Indexing.cu:1859]: Where is this from?

File: aten/src/ATen/native/cuda/Indexing.cu [hunk 4]
```diff
@@ -1833,7 +1846,41 @@ Tensor & masked_fill__cuda(Tensor& self, const Tensor & mask, const Tensor & val
...
   // exists in `masked_fill__cuda` (Scalar version).
   TORCH_CHECK(!self.device().is_cpu(), "masked_fill_: Expected inputs to be on same device")
-  return masked_fill__cuda(self, mask, value.item());
+  // A host-side value carries no device sync, so keep the cheap Scalar path.
+  if (value.device().is_cpu()) {
+    return masked_fill__cuda(self, mask, value.item());
+  }
+  TORCH_CHECK(self.device() == mask.device(), "expected self and mask to be on the same device, but got mask on ",
+    mask.device(), " and self on ", self.device());
+  TORCH_CHECK(mask.scalar_type() == kBool,
```
Comment (albanD) [inline aten/src/ATen/native/cuda/Indexing.cu:1855]: Isn't this already checked above?

File: aten/src/ATen/native/cuda/Indexing.cu [hunk 5]
```diff
@@ -1833,7 +1846,41 @@ Tensor & masked_fill__cuda(Tensor& self, const Tensor & mask, const Tensor & val
...
+    "masked_fill only supports boolean masks, but got dtype ", mask.scalar_type());
+  if (at::has_internal_overlap(self) == MemOverlap::Yes) {
+    TORCH_WARN(
+      "Use of masked_fill_ on expanded tensors is deprecated. "
+      "Please clone() the tensor before performing this operation. "
+      "This also applies to advanced indexing e.g. tensor[mask] = scalar");
+  }
+  at::assert_no_partial_overlap(self, mask);
```
Comment (albanD) [inline aten/src/ATen/native/cuda/Indexing.cu:1863]: There is no TensorIteartor flag for this?

File: test/test_torch.py [hunk 6]
```diff
@@ -3950,6 +3950,42 @@ def test_masked_fill_bool_tensor(self, device):
         dst = dst.masked_fill(mask, False)
         self.assertEqual(dst, torch.tensor([True, False, True], device=device))
 
+    def test_masked_fill_tensor_value(self, device):
+        # A 0-dim value tensor on the same device must match the Scalar overload.
+        dst = torch.arange(10, dtype=torch.float, device=device)
+        mask = (dst % 2 == 0)
+        ref = dst.clone().masked_fill_(mask, 3.5)
+        val = torch.tensor(3.5, device=device)
+        out = dst.clone().masked_fill_(mask, val)
```
Comment (albanD) [inline test/test_torch.py:3959]: Don't we have a context manager that raises on device sync? Should we use that here to assert there is no sync

File: aten/src/ATen/native/cuda/Indexing.cu (hunk 4 above)
Comment (dumko2001) [reply aten/src/ATen/native/cuda/Indexing.cu:1855]: The line above only guards self against being CPU; this checks self and mask share a device. This path doesn't route through the Scalar overload where that check normally lives.

File: aten/src/ATen/native/cuda/Indexing.cu (hunk 3 above)
Comment (dumko2001) [reply aten/src/ATen/native/cuda/Indexing.cu:1859]: From the Scalar overload (and masked_fill_impl_cpu). This path builds its own iterator instead of dispatching to the Scalar overload, so the mask-dtype guard isn't covered otherwise.

File: aten/src/ATen/native/cuda/Indexing.cu (hunk 5 above)
Comment (dumko2001) [reply aten/src/ATen/native/cuda/Indexing.cu:1863]: set_check_mem_overlap(true) errors on internal overlap, but masked_fill_ only warns there (deprecation) and errors on partial overlap with the mask. Kept the manual check to preserve that and match the Scalar overload.

File: test/test_torch.py (hunk 6 above)
Comment (dumko2001) [reply test/test_torch.py:3959]: Added test_masked_fill_tensor_value_no_host_sync using CudaSyncGuard("error") to assert no sync directly.

**Review by Skylion007 — APPROVED** [review]
LGTM

**Comment by dumko2001** [comment]
> LGTM

Thanks

**Review by albanD — APPROVED** [review]
Sounds good.
Thanks for the fix.

**Comment by dumko2001** [comment]
Hey @albanD, whats the next steps for me here ?
